src/main/java/address/sync/CloudSimulator.java

```diff
@@ -1,129 +1,670 @@
+    }
+
+    private void writeCloudAddressBookToFile(String addressBookName, CloudAddressBook cloudAddressBook) throws JAXBException {
+        File cloudFile = getCloudDataFilePath(addressBookName);
+        System.out.println("Writing to cloudFile: " + cloudFile.canRead());
+        try {
+            XmlFileHelper.saveCloudDataToFile(cloudFile, cloudAddressBook);
+        } catch (JAXBException e) {
+            System.out.println("Error writing to cloud file.");
+            throw e;
+        }
+    }
+
+    private void createCloudAddressBookFile(String addressBookName) throws IOException {
+        File cloudFile = getCloudDataFilePath(addressBookName);
+        if (cloudFile.exists()) {
+            throw new IllegalArgumentException("AddressBook '" + addressBookName + "' already exists!");
+        }
+        cloudFile.createNewFile();
+    }
+
+    private RawCloudResponse getEmptyResponse(int responseCode) {
+        return new RawCloudResponse(responseCode, null, convertToInputStream(getStandardHeaders()));
+    }
+
+    private void resetQuotaAndRestartTimer() {
+        long nextResetTime = getNextResetTime();
+        rateLimitStatus.setQuotaResetTime(nextResetTime);
+        rateLimitStatus.setQuotaRemaining(API_QUOTA_PER_HOUR);
+        if (timer != null) {
+            timer.stop();
+        }
+        int timeout = (int) (rateLimitStatus.getQuotaReset() - LocalDateTime.now().toEpochSecond(getSystemTimezone()));
+        timer = new TickingTimer("Cloud Quota Reset Time", timeout, this::printTimeLeft,
+                                this::resetQuotaAndRestartTimer, TimeUnit.SECONDS);
+        timer.start();
+    }
+
+    private void printTimeLeft(int timeLeft) {
+        System.out.println(timeLeft + " seconds remaining to quota reset.");
+    }
+
+    private List<CloudPerson> filterPersonsByTime(List<CloudPerson> personList, LocalDateTime time) {
+        return personList.stream()
+                .filter(person -> !person.getLastUpdatedAt().isBefore(time))
+                .collect(Collectors.toList());
+    }
+
+    private long getNextResetTime() {
+        LocalDateTime curTime = LocalDateTime.now();
+        LocalDateTime nearestHour = LocalDateTime.of(
+                curTime.getYear(), curTime.getMonth(), curTime.getDayOfMonth(), curTime.getHour() + 1,
+                0, 0, 0);
+
+        return nearestHour.toEpochSecond(getSystemTimezone());
+    }
+
+    private ZoneOffset getSystemTimezone() {
+        LocalDateTime localDateTime = LocalDateTime.now();
+        ZonedDateTime zonedDateTime = localDateTime.atZone(ZoneOffset.systemDefault());
+        return zonedDateTime.getOffset();
+    }
+
+    private boolean shouldSimulateNetworkFailure() {
+        return shouldSimulateUnreliableNetwork && RANDOM_GENERATOR.nextDouble() <= FAILURE_PROBABILITY;
+    }
+
+    private boolean shouldSimulateSlowResponse() {
+        return shouldSimulateUnreliableNetwork && RANDOM_GENERATOR.nextDouble() <= FAILURE_PROBABILITY;
```

Comment:
Maybe have another variable called `NETWORK_CONGESTION_PROBABILITY` or something, to differentiate with network failure. Also, this can have higher chance than failure.